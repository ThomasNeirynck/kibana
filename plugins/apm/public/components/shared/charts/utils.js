import numeral from '@elastic/numeral';

export function getYMaxRounded(yMax) {
  if (yMax <= 0) {
    return 0;
  }

  const yMaxUpper = yMax * 1.1;
  const initialBase = Math.floor(Math.log10(yMaxUpper));
  const base = initialBase > 2 ? initialBase - 1 : initialBase;
  return Math.ceil(yMaxUpper / 10 ** base) * 10 ** base;
}

export function getFormattedResponseTime(duration) {
  if (!duration && duration !== 0) {
    return `N/A`;
  }
  const formattedDuration = numeral(duration).format('0,0');
  return `${formattedDuration} ms`;
}

export function getFormattedRequestsPerMinute(rpm) {
  if (!rpm && rpm !== 0) {
    return `N/A`;
  }
  const formattedRpm = numeral(rpm).format('0.0');
  return `${formattedRpm} rpm`;
}
