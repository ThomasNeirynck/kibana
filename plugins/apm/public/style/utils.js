export function truncate(width) {
  return `
      max-width: ${width};
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    `;
}
