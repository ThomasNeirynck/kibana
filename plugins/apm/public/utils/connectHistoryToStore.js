import _ from 'lodash';
import { updateLocation } from '../store/location';

export default function connectHistoryToStore(history, dispatch) {
  function init() {
    const location = history.location;
    dispatch(updateLocation(location));
  }
  init();

  history.listen(
    _.debounce(location => {
      dispatch(updateLocation(location));
    }, 10)
  );
}
