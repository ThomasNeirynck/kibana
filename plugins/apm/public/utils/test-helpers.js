import configureStore from '../store/config/configureStore';
import PropTypes from 'prop-types';
import createHistory from 'history/createHashHistory';

export function getMockedStoreAndRouter() {
  const store = configureStore();
  const history = createHistory();

  return {
    context: { store, router: { history, route: { location: {} } } },
    childContextTypes: {
      store: PropTypes.object.isRequired,
      router: PropTypes.object.isRequired
    }
  };
}
