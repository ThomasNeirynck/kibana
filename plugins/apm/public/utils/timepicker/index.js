import moment from 'moment';
import { uiModules } from 'ui/modules';
import 'ui/chrome';
import 'ui/autoload/all';
import { debounce, set } from 'lodash';
import { setupRoutes } from '../../services/breadcrumbs';
import { legacyDecodeURIComponent, toQuery } from '../../utils/url';
import { updateTimePicker } from '../../store/urlParams';

const routes = {
  '/': 'APM',
  '/getting-started': 'Getting Started',
  '/:appName': {
    url: params => `/${params.appName}/transactions`,
    label: params => params.appName
  },
  '/:appName/errors': 'Errors',
  '/:appName/errors/:groupId': params => params.groupId,
  '/:appName/transactions': {
    skip: true
  },
  '/:appName/transactions/:transactionType': params => params.transactionType,
  '/:appName/transactions/:transactionType/:transactionName': params =>
    legacyDecodeURIComponent(params.transactionName)
};
const getBreadcrumbs = setupRoutes(routes);

let globalTimefilter;
let currentInterval;

export function initTimepicker(history, dispatch, callback) {
  uiModules.get('kibana').run(uiSettings => {
    set(
      uiSettings,
      'defaults.timepicker:timeDefaults.value',
      JSON.stringify({
        from: 'now-24h',
        to: 'now',
        mode: 'quick'
      })
    );
  });
  uiModules
    .get('app/apm', [])
    .controller('TimePickerController', ($scope, timefilter, globalState) => {
      $scope.breadcrumbs = getBreadcrumbs(history.location.pathname);
      $scope.searchQueryTime = toQuery(history.location.search)._g;

      history.listen(location => {
        updateRefreshRate(dispatch, timefilter);

        $scope.$apply(() => {
          $scope.breadcrumbs = getBreadcrumbs(location.pathname);
          $scope.searchQueryTime = toQuery(history.location.search)._g;
        });
        globalState.fetch();
      });
      timefilter.setTime = (from, to) => {
        timefilter.time.from = moment(from).toISOString();
        timefilter.time.to = moment(to).toISOString();
        $scope.$apply();
      };
      timefilter.enabled = true;
      timefilter.init();

      updateRefreshRate(dispatch, timefilter);

      timefilter.on(
        'update',
        debounce(() => dispatch(getAction(timefilter)), 10)
      );

      // hack to access timefilter outside Angular
      globalTimefilter = timefilter;
      callback();
    });
}

function getAction(timefilter) {
  return updateTimePicker({
    min: timefilter.getBounds().min.toISOString(),
    max: timefilter.getBounds().max.toISOString()
  });
}

function updateRefreshRate(dispatch, timefilter) {
  const refreshInterval = timefilter.refreshInterval.value;
  if (currentInterval) {
    clearInterval(currentInterval);
  }

  if (refreshInterval > 0 && !timefilter.refreshInterval.pause) {
    currentInterval = setInterval(
      () => dispatch(getAction(timefilter)),
      refreshInterval
    );
  }
}

export function getTimefilter() {
  if (!globalTimefilter) {
    throw new Error(
      'Timepicker must be initialized before calling getTimefilter'
    );
  }
  return globalTimefilter;
}
