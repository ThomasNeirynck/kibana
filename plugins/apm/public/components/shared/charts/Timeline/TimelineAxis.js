import React from 'react';
import _ from 'lodash';
import { Sticky } from 'react-sticky';
import { XYPlot, XAxis } from 'react-vis';
import LastTickValue from './LastTickValue';
import { getTimeFormatter } from '../../../../utils/formatters';
import { colors } from '../../../../style/variables';

// Remove last tick if it's too close to xMax
const getXAxisTickValues = (tickValues, xMax) =>
  _.last(tickValues) * 1.05 > xMax ? tickValues.slice(0, -1) : tickValues;

function TimelineAxis({
  xScale,
  xDomain,
  width,
  timelineMargins,
  tickValues,
  xMax,
  header
}) {
  const tickFormat = getTimeFormatter(xMax);
  const xAxisTickValues = getXAxisTickValues(tickValues, xMax);

  return (
    <Sticky disableCompensation>
      {({ style }) => {
        return (
          <div
            style={{
              position: 'absolute',
              backgroundColor: colors.white,
              borderBottom: `1px solid ${colors.gray3}`,
              height: `${timelineMargins.top}px`,
              display: 'flex',
              justifyContent: 'space-between',
              flexDirection: 'column',
              zIndex: 2,
              ...style
            }}
          >
            {header}
            <XYPlot
              dontCheckIfEmpty
              width={width}
              height={20}
              margin={{
                top: 20,
                left: timelineMargins.left,
                right: timelineMargins.right
              }}
              xDomain={xDomain}
            >
              <XAxis
                hideLine
                orientation="top"
                tickSize={0}
                tickValues={xAxisTickValues}
                tickFormat={tickFormat}
                style={{
                  text: { fill: colors.gray3 }
                }}
              />

              <LastTickValue x={xScale(xMax)} value={tickFormat(xMax)} />
            </XYPlot>
          </div>
        );
      }}
    </Sticky>
  );
}

export default TimelineAxis;
