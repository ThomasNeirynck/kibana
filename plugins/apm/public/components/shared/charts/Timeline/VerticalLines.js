import React, { PureComponent } from 'react';
import { XYPlot, VerticalGridLines } from 'react-vis';
import { colors } from '../../../../style/variables';

export default class VerticalLines extends PureComponent {
  render() {
    const {
      width,
      height,
      timelineMargins,
      xDomain,
      tickValues,
      xMax
    } = this.props;

    return (
      <div
        style={{
          position: 'absolute',
          top: 0,
          left: 0
        }}
      >
        <XYPlot
          dontCheckIfEmpty
          width={width}
          height={height + timelineMargins.top}
          margin={timelineMargins}
          xDomain={xDomain}
        >
          <VerticalGridLines
            tickValues={tickValues}
            style={{ stroke: colors.gray5 }}
          />
          <VerticalGridLines
            tickValues={[xMax]}
            style={{ stroke: colors.gray3 }}
          />
        </XYPlot>
      </div>
    );
  }
}
