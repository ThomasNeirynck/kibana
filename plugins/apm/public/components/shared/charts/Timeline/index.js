import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import 'react-vis/dist/style.css';
import { scaleLinear } from 'd3-scale';
import { makeWidthFlexible } from 'react-vis';
import TimelineAxis from './TimelineAxis';
import VerticalLines from './VerticalLines';

const getXScale = _.memoize(
  (xMin, xMax, timelineMargins, width) => {
    return scaleLinear()
      .domain([xMin, xMax])
      .range([timelineMargins.left, width - timelineMargins.right]);
  },
  (xMin, xMax, timelineMargins, width) =>
    [xMin, xMax, timelineMargins.left, timelineMargins.right, width].join('__')
);

const getTicks = _.memoize(
  xScale => xScale.ticks(7),
  xScale => xScale.domain().join('__')
);
const getXDomain = _.memoize(
  xScale => xScale.domain(),
  xScale => xScale.domain().join('__')
);

class Timeline extends PureComponent {
  render() {
    const { width, height, timelineMargins, duration, header } = this.props;

    if (duration == null || !width) {
      return null;
    }

    const xMin = 0;
    const xMax = duration;
    const xScale = getXScale(xMin, xMax, timelineMargins, width);
    const xDomain = getXDomain(xScale);
    const tickValues = getTicks(xScale);

    return (
      <div>
        <TimelineAxis
          width={width}
          timelineMargins={timelineMargins}
          xScale={xScale}
          xDomain={xDomain}
          tickValues={tickValues}
          xMax={xMax}
          header={header}
        />

        <VerticalLines
          width={width}
          height={height}
          timelineMargins={timelineMargins}
          xDomain={xDomain}
          tickValues={tickValues}
          xMax={xMax}
        />
      </div>
    );
  }
}

Timeline.propTypes = {
  width: PropTypes.number
};

export default makeWidthFlexible(Timeline);
