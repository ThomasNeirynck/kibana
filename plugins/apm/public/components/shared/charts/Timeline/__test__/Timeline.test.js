import React from 'react';
import { mount } from 'enzyme';
import { StickyContainer } from 'react-sticky';

import Timeline from '../index';
import props from './props.json';
import { mockMoment, toJson } from '../../../../../utils/testHelpers';

describe('Timline', () => {
  beforeAll(() => {
    mockMoment();
  });

  it('should render with data', () => {
    const wrapper = mount(
      <StickyContainer>
        <Timeline header={<div>Hello - i am a header</div>} {...props} />
      </StickyContainer>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
