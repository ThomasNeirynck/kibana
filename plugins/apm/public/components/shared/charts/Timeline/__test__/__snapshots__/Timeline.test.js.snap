// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timline should render with data 1`] = `
"
<div>
  <div style=\\"width: 100%; height: 100%;\\">
    <div>
      <div>
        <div>
        </div>
        <div style=\\"position: absolute; background-color: rgb(255, 255, 255); border-bottom: 1px solid #999999; height: 100px; z-index: 2; width: 100%;\\">
          <div>
            Hello - i am a header
          </div>
          <div style=\\"width: 100px; height: 40px;\\"
               class=\\"rv-xy-plot \\"
          >
            <svg class=\\"rv-xy-plot__inner\\"
                 width=\\"100\\"
                 height=\\"40\\"
            >
              <g transform=\\"translate(50,0)\\"
                 class=\\"rv-xy-plot__axis rv-xy-plot__axis--horizontal \\"
              >
                <g transform=\\"translate(0, 40)\\"
                   class=\\"rv-xy-plot__axis__ticks\\"
                >
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      0
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      20 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      40 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      60 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      80 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      100 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      120 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      140 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      160 ms
                    </text>
                  </g>
                  <g transform=\\"translate(0, 0)\\"
                     class=\\"rv-xy-plot__axis__tick\\"
                  >
                    <line x1=\\"0\\"
                          x2=\\"0\\"
                          y1=\\"0\\"
                          y2=\\"0\\"
                          class=\\"rv-xy-plot__axis__tick__line\\"
                    >
                    </line>
                    <text text-anchor=\\"middle\\"
                          dy=\\"0\\"
                          transform=\\"translate(0, -8)\\"
                          class=\\"rv-xy-plot__axis__tick__text\\"
                          style=\\"fill: #999999;\\"
                    >
                      180 ms
                    </text>
                  </g>
                </g>
              </g>
              <g transform=\\"translate(50, 40)\\">
                <text text-anchor=\\"middle\\"
                      dy=\\"0\\"
                      transform=\\"translate(0, -8)\\"
                >
                  205 ms
                </text>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div style=\\"position: absolute; top: 0px; left: 0px;\\">
        <div style=\\"width: 100px; height: 216px;\\"
             class=\\"rv-xy-plot \\"
        >
          <svg class=\\"rv-xy-plot__inner\\"
               width=\\"100\\"
               height=\\"216\\"
          >
            <g transform=\\"translate(50,100)\\"
               class=\\"rv-xy-plot__grid-lines\\"
            >
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #f5f5f5;\\"
              >
              </line>
            </g>
            <g transform=\\"translate(50,100)\\"
               class=\\"rv-xy-plot__grid-lines\\"
            >
              <line x1=\\"0\\"
                    x2=\\"0\\"
                    y1=\\"0\\"
                    y2=\\"116\\"
                    class=\\"rv-xy-plot__grid-lines__line\\"
                    style=\\"stroke: #999999;\\"
              >
              </line>
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
"
`;
